### PR TITLE
Introduce IDisposable to ArrayBuilder

### PR DESCRIPTION
**Pull request by petrroll**
Introduced IDisposable to ArrayBuilder to enable `using ….` pattern. ~Already found one leak. Only 2296d68 is new commit, the rest is from #37295~ Rebased.

### CODE REVIEW

**Comment by mavasani** [comment]
@petrroll Can you rebase your source branch so it is based off master instead of https://github.com/dotnet/roslyn/pull/37295? That would make it easier for reviewing for the compiler team, as this is mostly a compiler layer change.

**Comment by jcouv** [comment]
>         public ImmutableArray<T> ToImmutableAndFree()

are there still uses of this method?

---
Refers to: src/Dependencies/PooledObjects/ArrayBuilder.cs:278 in 2296d68. [](commit_id = 2296d6867190f261fa1ade98b82d145beae67717, deletion_comment = False)

File: src/EditorFeatures/CSharpTest/CodeActions/AddAwait/AddAwaitTests.cs [hunk 1]
```diff
@@ -259,19 +260,75 @@ async Task<int> GetNumberAsync()
         }
 
         [Fact]
-        public async Task MissingOnSemiColon()
+        [WorkItem(35525, "https://github.com/dotnet/roslyn/issues/35525")]
+        public async Task OnSemiColon()
         {
-            await TestMissingInRegularAndScriptAsync(@"
+            await TestInRegularAndScriptAsync(@"
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/CodeActions/AddAwait/AddAwaitTests.cs]: Are these IDE changes intended in this PR? Seems unrelated

**Comment by petrroll** [comment]
Rebased.

File: src/EditorFeatures/CSharpTest/CodeActions/AddAwait/AddAwaitTests.cs (hunk 1 above)
Comment (petrroll) [reply src/EditorFeatures/CSharpTest/CodeActions/AddAwait/AddAwaitTests.cs]: It was based on a different PR, rebased to master for clarity.

Comment (petrroll) [reply src/EditorFeatures/CSharpTest/CodeActions/AddAwait/AddAwaitTests.cs]: Didn't rebased correctly. WIP

File: src/Compilers/Core/Portable/InternalUtilities/EnumerableExtensions.cs [hunk 2]
```diff
@@ -501,6 +501,7 @@ internal static class Functions<T>
     {
         public static readonly Func<T, T> Identity = t => t;
         public static readonly Func<T, bool> True = t => true;
+        public static readonly Func<T, bool> False = t => false;
```
Comment (jcouv) [inline src/Compilers/Core/Portable/InternalUtilities/EnumerableExtensions.cs]: is this used?

**Comment by petrroll** [comment]
Issues with rebasing, will fix & reopen.

**Comment by petrroll** [comment]
Superseded by #37543
(easier than force pushing & fixing this)
